python/ray/data/datasource/sql_datasource.py

```diff
@@ -0,0 +1,175 @@
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        # If `parallelism` is 1, directly read the entire database and avoid queries
+        # to fetch a sample block and compute the total number of rows.
+        if parallelism == 1:
+
+            def read_fn() -> Iterable[Block]:
+                with connect(self.connection_factory) as cursor:
+                    cursor.execute(f"SELECT * FROM ({self.sql}) as T")
+                    block = cursor_to_block(cursor)
+                    return [block]
+
+            metadata = BlockMetadata(None, None, None, None, None)
+            return [ReadTask(read_fn, metadata)]
+
+        num_rows_total = self._get_num_rows()
+
+        if num_rows_total == 0:
+            return []
+
+        parallelism = min(
+            parallelism, math.ceil(num_rows_total / self.MIN_ROWS_PER_READ_TASK)
+        )
+        num_rows_per_block = num_rows_total // parallelism
+        num_blocks_with_extra_row = num_rows_total % parallelism
+
+        sample_block_accessor = BlockAccessor.for_block(self._get_sample_block())
+        estimated_size_bytes_per_row = math.ceil(
+            sample_block_accessor.size_bytes() / sample_block_accessor.num_rows()
+        )
+        sample_block_schema = sample_block_accessor.schema()
+
+        tasks = []
+        offset = 0
+        for i in range(parallelism):
+            num_rows = num_rows_per_block
+            if i < num_blocks_with_extra_row:
+                num_rows += 1
+
+            read_fn = self._create_read_fn(num_rows, offset)
+            metadata = BlockMetadata(
+                num_rows,
+                estimated_size_bytes_per_row * num_rows,
+                sample_block_schema,
+                None,
+                None,
+            )
+            tasks.append(ReadTask(read_fn, metadata))
+
+            offset += num_rows
+
+        return tasks
+
+    def _get_num_rows(self) -> int:
+        with connect(self.connection_factory) as cursor:
+            cursor.execute(f"SELECT COUNT(*) FROM ({self.sql}) as T")
+            return cursor.fetchone()[0]
+
+    def _get_sample_block(self) -> Block:
+        with connect(self.connection_factory) as cursor:
+            cursor.execute(
+                f"SELECT * FROM ({self.sql}) as T LIMIT {self.NUM_SAMPLE_ROWS}"
+            )
+            return cursor_to_block(cursor)
+
+    def _create_read_fn(self, num_rows: int, offset: int):
+        def read_fn() -> Iterable[Block]:
+            with connect(self.connection_factory) as cursor:
+                try:
+                    query = (
+                        f"SELECT * FROM ({self.sql}) as T LIMIT {num_rows} "
+                        f"OFFSET {offset}"
+                    )
+                    cursor.execute(query)
+                except Exception as e:
+                    # Some databases don't support `LIMIT` and `OFFSET`.
+                    raise RuntimeError("Failed to execute SQL query '{query!r}'") from e
```

Comment:
Instead of throwing exception, we should have fallback to launch 1 read task instead (same as parallelism=1). Users do not specify that they want to do `LIMIT/OFFSET` thing, and this is our internal logic, so it's unfair to throw exception to them.

Reply:
I agree that we aren't gracefully handling the error here.

Do you know how we might be able to implement this? By the time one of the read tasks fails, `get_read_tasks` would've returned, and the read tasks would've already been launched. I'm not sure how we would catch the error and re-submit a single read task.

One workaround is to execute a query at the start of `get_read_tasks` to check if LIMIT is supported like:
```python
        # Databases like DB2, Oracle, and MS SQL Server don't support `LIMIT`.
        try:
            with connect(self.connection_factory) as cursor:
                cursor.execute(f"SELECT * FROM ({self.sql}) as T LIMIT 1 OFFSET 0")
            is_limit_supported = True
        except Exception:
            is_limit_supported = False
```

The problem with this approach is that it may introduce performance overhead.